### PR TITLE
Roll coords deprecation

### PR DESCRIPTION
**Pull request by TomNicholas**
The default behaviour of `da.roll()` caught me out whilst trying to hand-write a `diff` function, so I completed the transition to defaulting to `roll_coords=False` as the default. It's been throwing a warning for 3 years so I think it's time!

I also improved the docstrings and added type hints whilst there, although mypy doesn't seem to like some of the type hinting :/

- [x] Completes deprecation started in #2360
- [x] Tests updated
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

**Comment by max-sixty** [comment]
@TomNicholas typing should be fixed!

**Comment by TomNicholas** [comment]
Thank you very much @max-sixty and @mathause ! I think this should be ready to merge now.

File: xarray/core/indexes.py [hunk 1]
```diff
@@ -440,7 +441,7 @@ class Indexes(collections.abc.Mapping):
 
     __slots__ = ("_indexes",)
 
-    def __init__(self, indexes):
+    def __init__(self, indexes: Mapping[Any, pd.Index]) -> None:
```
Comment (mathause) [inline xarray/core/indexes.py:444]: @benbovy `Indexes` is typed here - is this compatible with the refactor (i.e. the `pd.Index`)? And if not can you adapt it in #5636?

Comment (benbovy) [reply xarray/core/indexes.py:444]: Ah yes I forgot to update the type here. `Indexes` is returned by both `Dataset.indexes` (pandas indexes) and `Dataset.xindexes` (xarray indexes), so `Union[pd.Index, Index]` is better.

Comment (dcherian) [reply xarray/core/indexes.py:444]:
```suggestion
    def __init__(self, indexes: Mapping[Any, Union[pd.Index, Index]]) -> None:
```

**Comment by max-sixty** [comment]
@TomNicholas shall we merge?

**Pull request merged**